cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -66,12 +77,47 @@ public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
 
       //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
       return writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
-    } finally {
-      try {
-        repositoryManager.close();
-      } catch (IOException e) {
-        LOG.warn("Failed to close the RepositoryManager, there may be leftover files", e);
+    }
+  }
+
+  @Override
+  public PullAppResponse pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws PullFailureException,
+    NotFoundException, AuthenticationConfigException, IllegalArgumentException {
+    String applicationName = appRef.getApplication();
+    String configFileName = generateConfigFileName(applicationName);
+    LOG.info("Cloning remote to pull application {}", applicationName);
+    String contents;
+    String fileHash;
+    Path appRelativePath;
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(appRef.getNamespaceId(), repoConfig)) {
+      appRelativePath = repositoryManager.getFileRelativePath(configFileName);
+      String commitID = repositoryManager.cloneRemote();
+      Path filePathToRead;
+      filePathToRead = validateAppConfigRelativePath(repositoryManager, appRelativePath);
+      if (!Files.exists(filePathToRead)) {
+        throw new NotFoundException(String.format("App with name %s not found in git repository", applicationName));
```

Comment:
Please include relative path. This may help in cases where customer expects it under a different name for whatever reason (prefixes, etc)

Reply:
Added.